stable_mir: Add `MutMirVisitor`

makai410 opened the pull request:
Resolves: [rust-lang/project-stable-mir#81](https://github.com/rust-lang/project-stable-mir/issues/81).

I am unsure if we should add a `MutableBody` like Kani did.
Currently, I use `&mut Body` in `MutMirVisitor::visit_body()`.

r? @celinval

celinval commented:
> Resolves: [rust-lang/project-stable-mir#81](https://github.com/rust-lang/project-stable-mir/issues/81).
> 
> I am unsure if we should add a `MutableBody` like Kani did. Currently, I use `&mut Body` in `MutMirVisitor::visit_body()`.

I like the overall solution. I'll review the implementation when I'm in my laptop. Thanks!

celinval commented on compiler/stable_mir/src/mir/body.rs:
```diff
@@ -77,6 +77,22 @@ impl Body {
         &self.locals[self.arg_count + 1..]
     }
 
+    /// Returns a mutable reference to the local that holds this function's return value.
+    pub fn ret_local_mut(&mut self) -> &mut LocalDecl {
+        &mut self.locals[RETURN_LOCAL]
+    }
+
+    /// Returns a mutable slice of locals corresponding to this function's arguments.
+    pub fn arg_locals_mut(&mut self) -> &mut [LocalDecl] {
+        &mut self.locals[1..][..self.arg_count]
+    }
+
+    /// Returns a mutable slice of inner locals for this function.
+    /// Inner locals are those that are neither the return local nor the argument locals.
+    pub fn inner_locals_mut(&mut self) -> &mut [LocalDecl] {
+        &mut self.locals[self.arg_count + 1..]
+    }
+
```
Can we keep these `pub(crate)` for now?

makai410 replied:
Done

celinval reviewed (approved):
Looks good! Thanks! Please r=me after you change the visibility of the new body methods.

makai410 committed:
Add `MutMirVisitor`
Implement `make_mir_visitor` macro to generate `MirVisitor` and `MutMirVisitor`.

Add `ret_local_mut()`, `arg_locals_mut()` and `inner_locals_mut()` to `Body`, specifically for `MutMirVisitor`.

makai410 committed:
Add test for `MutMirVisitor`

makai410 commented:
@bors r=celinval

bors commented:
@makai410: :key: Insufficient privileges: Not in reviewers

makai410 commented:
> Looks good! Thanks! Please r=me after you change the visibility of the new body methods.

Sorry that I am not a team member actually😥

celinval commented:
@bors r+ rollup

I didn't know you needed one.... Thanks again!

bors commented:
:pushpin: Commit ad315f6074868024689aaa98021bffdd590e0841 has been approved by `celinval`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=celinval ad315f6074868024689aaa98021bffdd590e0841 -->
<!-- homu: {"type":"Approved","sha":"ad315f6074868024689aaa98021bffdd590e0841","approver":"celinval","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
